File: android/app/src/main/java/com/petclump/petclump/models/OwnerProfile.java
package com.petclump.petclump.models;

import android.content.Context;
import android.location.Location;
import android.location.LocationManager;
import android.util.Log;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.petclump.petclump.models.protocols.Profile;
import com.petclump.petclump.models.protocols.ProfileDownloader;
import com.petclump.petclump.models.protocols.ProfileUploader;

import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;



/**
 * This is the onwer profile data model
 * - id: this should be the uid from Firebase Auth object
 * - name: this should be user's name to show to public
 * - birthday: this is a Date object. In the initlizer, it's a yyyy/MM/dd string
 * - gender: a string of any gender
 * - distancePerference: this shows the user's perference for matching
 * - lat and lon: for location calculation
 * - freeTime: this is a 7*3 bool matrix for free time. In the initlizer, it's a 21 character string with 1 mark as free.
 */
public class OwnerProfile implements Profile {
    /*** data field ***/
    private String name     = "No name";
    private String gender   = "Apache";
    private Date birthday   = new Date();
    private int distancePerference = 5;
    private double lat = 0.0 ,lon = 0.0;
    private FreeSchedule freeTime = new FreeSchedule("");

    private String TAG = "Ownerprofile";

    // singleton
    private static OwnerProfile obj = new OwnerProfile();
    private OwnerProfile (){}


    public static OwnerProfile getInstance(){
        return obj;
    }

    public static String num_month(int num){
        switch(num){
            case 1: return "Jan";
            case 2: return "Feb";
            case 3: return "Mar";
            case 4: return "Apr";
            case 5: return "May";
            case 6: return "Jun";
            case 7: return "Jul";
            case 8: return "Aug";
            case 9: return "Sep";
            case 10: return "Oct";
            case 11: return "Nov";
            case 12: return "Dec";
        }
        return "month_error "+num;
    }
    @Override
    public Map<String,Object> generateDictionary(){
        return  new HashMap<String,Object>(){{
            put("lat", lat);
            put("lon", lon);
            put("name", name);
            put("gender", gender);
            put("birthday", birthday);
            put("freeTime", freeTime.freeString);
            put("distancePerference", distancePerference);
        }};

    }
    @Override
    public void upload(String id, ProfileUploader c){
        if (FirebaseAuth.getInstance().getCurrentUser() == null){
            Log.d(TAG, "upload_failed due to null user! ");
        }
        FirebaseFirestore.getInstance().collection("users").document(id).set(generateDictionary()).addOnCompleteListener(task -> {
            //Log.d("Profile", "upload: " + message);
            c.didCompleteUpload();
        });
    }
    @Override
    public void download(String id, ProfileDownloader c){
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        String TAG = "OwnerProfile_"+name;
        if (user == null){ Log.w(TAG,"empty user");  return; }
        FirebaseFirestore.getInstance().collection("users").document(id).get().addOnCompleteListener(task->{
            if (!task.isSuccessful()) {
                Log.d(TAG, "Download failed: " + user.getUid());
                return;
            }

            Map<String, Object> ref = task.getResult().getData();
            this.name = (String) ref.get("name");
            Log.d(TAG, "Download successfully" + name);
            this.gender = (String) ref.get("gender");
            Log.d(TAG, "Download successfully" + gender);
            this.birthday = (Date) ref.get("birthday");
            Log.d(TAG, "Download successfully" + String.valueOf(birthday));
            this.distancePerference = Integer.parseInt(ref.get("distancePerference").toString());
            this.lat = (Double)ref.get("lat");
            this.lon = (Double)ref.get("lon");
            this.freeTime = new FreeSchedule((String) ref.get("freeTime"));

            c.didCompleteDownload();
        });


    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public Date getBirthday() {
        return birthday;
    }

    public void setBirthday(Date birthday) {
        this.birthday = birthday;
    }

    public int getDistancePerference() {
        return distancePerference;
    }

    public void setDistancePerference(int distancePerference) {
        this.distancePerference = distancePerference;
    }


    public double getLat() {
        return lat;
    }
    public double getLon() {
        return lon;
    }
    public void setLat(double lat) {
        this.lat = lat;
    }
    public void setLon(double lon) {
        this.lon = lon;
    }

    public FreeSchedule getFreeTime() {
        return freeTime;
    }

    public void setFreeTime(String freeTime) {
        this.freeTime = new FreeSchedule(freeTime);
    }
}

File: ios/PetClump/Model/OwnerProfile.swift
//
//  OwnerProfile.swift
//  PetClump
//
//  Created by YSH on 4/29/18.
//  Copyright © 2018 PetClump. All rights reserved.
//

import Foundation
import CoreLocation
import Firebase

/**
 * This is the onwer profile data model
 * - id: this should be the uid from Firebase Auth object
 * - name: this should be user's name to show to public
 * - birthday: this is a Date object. In the initlizer, it's a yyyy/MM/dd string
 * - gender: a string of any gender
 * - distancePerference: this shows the user's perference for matching
 * - lat and lon: for location calculation
 * - freeTime: this is a 7*3 bool matrix for free time. In the initlizer, it's a 21 character string with 1 mark as free.
 */
class OwnerProfile: Profile{
    func upload(vc: QuickAlert, completion: ProfileUploader?) {
        guard let uid = Auth.auth().currentUser?.uid else { return }
        
        let docRef = Firestore.firestore().collection("users").document(uid)
        docRef.setData(self.generateDictionary()) { (err: Error?) in
            if let err = err{
                vc.makeAlert(message: "Upload failed, reason:" + err.localizedDescription)
            }
            print("Uploaded successfully for user " + uid)
            guard (completion != nil) else { return }
            completion!.didCompleteUpload()
        }
    }
    
    func download(uid: String, completion: ProfileDownloader?) {
        // Opens document
        let docRef =  Firestore.firestore().collection(COLLECTION_NAME).document(uid)
        docRef.getDocument { (document, error) in
            if let document = document, document.exists {
                // Unwraps data object
                let refObj = document.data()!
                print("Document data: \(refObj.description)")
                
                // Gets user information
                self.name = refObj["name"] as? String ?? ""
                self.gender = refObj["gender"] as? String ?? ""
                
                // Gets user birthdat and parse it
                if let bd = refObj["birthday"] as? Timestamp{
                    self.birthday = bd.dateValue()
                }
                
                // Gets match perference and updates the slider
                self.distancePerference = refObj["distancePerference"] as? Int ?? 25
                
                // Gets Freetime and convert to Free schedule
                self.freeTime = FreeSchedule(freeString: refObj["freeTime"] as? String ?? "")
            }
            guard (completion != nil) else { return }
            completion!.didCompleteDownload()
        }
    }
    
    
    private let COLLECTION_NAME:String = "users"
    var name: String = "No name"
    var birthday: Date = Date()
    var gender: String = "Apache"
    var distancePerference: Int = 5
    var lat: Double = 0.0
    var lon: Double = 0.0
    var freeTime = FreeSchedule(freeString: "")

    
    func generateDictionary() -> [String : Any] {
        return [
            "lat":  lat ,
            "lon":  lon ,
            "name": name ,
            "gender":   gender ,
            "birthday": birthday ,
            "freeTime": freeTime.freeTimeAsString,
            "distancePerference": distancePerference
        ]
    }
    
    func getBirthdayString() -> String{
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy/MM/dd"
        return dateFormatter.string(from: self.birthday)
    }
}


class FreeSchedule{
    var freeMatrix = Array(repeating: Array(repeating: false, count: 3), count:7)
    var freeTimeAsString: String
    
    init(freeString: String){
        self.freeTimeAsString = freeString
        var chars = Array(freeString), manCounter = 0, weekCounter = 0
        if chars.count != 21 {
            chars = Array("000000000000000000000")
        }
        for char in chars {
            // 0 is not free, 1 is free
            self.freeMatrix[weekCounter][manCounter] = char == "1"
            manCounter += 1
            manCounter %= 3
            if manCounter == 0 { weekCounter += 1 }
        }
    }
    
    func isFree(weekDay: Int, partDay: Int) -> Bool {
        return freeMatrix[weekDay][partDay]
    }
    
    func getCommonFreeTime(other: FreeSchedule) -> FreeSchedule{
        var commonStirng = ""
        let thisTime = Array(self.freeTimeAsString)
        let otherTime = Array(other.freeTimeAsString)
        for index in 0...thisTime.count {
            if thisTime[index] == "1" && otherTime[index] == "1"{
                commonStirng += "1"
            } else {
                commonStirng += "0"
            }
        }
        return FreeSchedule(freeString: commonStirng)
    }
}

